src/bp/nlu-server/api.ts

```diff
@@ -0,0 +1,140 @@
+import { NLUServerLogger } from './logger'
+import ModelService from './model-service'
+import { monitoringMiddleware, startMonitoring } from './monitoring'
+import { authMiddleware, handleErrorLogging, handleUnexpectedError } from './util'
+import { TrainInput, TrainInputCreateSchema } from './validation'
+
+export type APIOptions = {
+  version: string
+  host: string
+  port: number
+  modelDir: string
+  authToken?: string
+  limitWindow: string
+  limit: number
+  adminToken: string
+}
+
+const debug = DEBUG('api')
+const debugRequest = debug.sub('request')
+const cachePolicy = { 'Cache-Control': `max-age=${ms('1d')}` }
+
+const createExpressApp = (options: APIOptions): Application => {
+  const app = express()
+
+  // This must be first, otherwise the /info endpoint can't be called when token is used
+  app.use(cors())
+
+  app.use(bodyParser.json({ limit: '250kb' }))
+
+  app.use((req, res, next) => {
+    res.header('X-Powered-By', 'Botpress')
+    debugRequest('incoming ' + req.path, { ip: req.ip })
+    next()
+  })
+
+  app.use(monitoringMiddleware)
+  app.use(handleUnexpectedError)
+
+  if (process.core_env.REVERSE_PROXY) {
+    app.set('trust proxy', process.core_env.REVERSE_PROXY)
+  }
+
+  if (options.limit > 0) {
+    app.use(
+      rateLimit({
+        windowMs: ms(options.limitWindow),
+        max: options.limit,
+        message: 'Too many requests, please slow down'
+      })
+    )
+  }
+
+  if (options.authToken && options.authToken.length) {
+    // Both tokens can be used to query the language server
+    app.use(authMiddleware(options.authToken, options.adminToken))
+  }
+
+  return app
+}
+
+export default async function(options: APIOptions) {
+  const app = createExpressApp(options)
+  const logger = new NLUServerLogger('API')
+  const loggerWrapper: NLU.Logger = {
+    info: (msg: string) => logger.info(msg),
+    warning: (msg: string, err?: Error) => (err ? logger.attachError(err).warn(msg) : logger.warn(msg)),
+    error: (msg: string, err?: Error) => (err ? logger.attachError(err).error(msg) : logger.error(msg))
+  }
+  const engine = new Engine('nlu-server', loggerWrapper)
+  const nluGhost = new NLUServerGhost()
+  const modelService = new ModelService(nluGhost, options.modelDir)
+  await modelService.createModelDirIfNotExist()
+
+  const adminTokenMw = authMiddleware(options.adminToken)
+
+  app.get('/info', (req, res) => {
+    res.send({
+      version: options.version
+    })
+  })
+
+  const router = express.Router({ mergeParams: true })
+  router.post('/train', async (req, res) => {
+    const input: TrainInput = await validate(req.body, TrainInputCreateSchema, {
+      stripUnknown: true
+    })
+
+    const intents = _.flatMap(Object.values(input.topics))
+    const model = await engine.train(intents, input.entities, input.language)
+
+    if (!model) {
+      res.sendStatus(500) // an error occured
+    }
+    const modelId = await modelService.saveModel(model!)
+    return res.send({
```

Comment:
I know this is a WIP but we don't want to wait until training done until we return the modelId.  We would have a GET route that exposes the training status from modelId (maybe add a TODO in the code here)